From 7e45c8c5bdcfb1dad9cee2249eade90758abe860 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Wed, 24 Aug 2022 11:48:31 +0300
Subject: [PATCH] [SW-81037] boot_desc: init descriptors once

boot_desc_init() will be called multiple times from different places
of the code, for example once for the PIF ARC images and later on for
the Preboot image. there is no need to read the descriptors more than
once, the flash content is not expected to change on runtime.

Change-Id: I23adb37e8e77f15201936daa3913e03d64ad6e83
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/lib/common/boot_desc.c b/lib/common/boot_desc.c
index cc8f904..e4550ba 100644
--- a/lib/common/boot_desc.c
+++ b/lib/common/boot_desc.c
@@ -23,6 +23,7 @@
 
 static struct hl_fw_boot_ptrs boot_ptrs;
 static struct hl_fw_img_desc img_desc[BOOT_DESC_MAX_IMGS];
+static bool desc_initialized;
 
 static void boot_desc_print_magic(char *name, struct hl_fw_boot_magic *magic)
 {
@@ -64,6 +65,9 @@
 	uint32_t copy_sz;
 	unsigned int i;
 
+	if (desc_initialized)
+		return boot_ptrs.count;
+
 	/* Read the main boot pointers structure */
 	if (ppboot_boot_info_get(HL_BOOT_INFO_IMG_DESC_INFO, &desc_info,
 				  sizeof(struct hl_fw_desc_info), &copy_sz)) {
@@ -92,6 +96,8 @@
 		}
 	}
 
+	desc_initialized = true;
+
 	return boot_ptrs.count;
 }
 
@@ -102,6 +108,9 @@
 	unsigned int count, i;
 	u32 spi_off = HL_FW_BOOT_PTRS_BASE;
 	u32 desc_offs[BOOT_DESC_MAX_IMGS] = { 0 };
+
+	if (desc_initialized)
+		return boot_ptrs.count;
 
 	/* Read the main boot pointers structure */
 	spi_flash_copy(&boot_ptrs, spi_off, BOOT_PTRS_SZ, "BOOT PTRS");
@@ -126,6 +135,8 @@
 		}
 	}
 
+	desc_initialized = true;
+
 	return count;
 }
 
